libs/pubsub/query/query.go

```diff
@@ -1,527 +1,327 @@
-// Package query provides a parser for a custom query format:
+// Package query implements the custom query format used to filter event
+// subscriptions in Tendermint.
 //
-//		abci.invoice.number=22 AND abci.invoice.owner=Ivan
+// Query expressions describe properties of events and their attributes, using
+// strings like:
 //
-// See query.peg for the grammar, which is a https://en.wikipedia.org/wiki/Parsing_expression_grammar.
-// More: https://github.com/PhilippeSigaud/Pegged/wiki/PEG-Basics
+//    abci.invoice.number = 22 AND abci.invoice.owner = 'Ivan'
+//
+// Query expressions can handle attribute values encoding numbers, strings,
+// dates, and timestamps.  The complete query grammar is described in the
+// query/syntax package.
 //
-// It has a support for numbers (integer and floating point), dates and times.
 package query
 
 import (
 	"fmt"
-	"reflect"
 	"regexp"
 	"strconv"
 	"strings"
 	"time"
 
 	"github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/pubsub/query/syntax"
 )
 
-var (
-	numRegex = regexp.MustCompile(`([0-9\.]+)`)
-)
+// All is a query that matches all events.
+var All *Query
 
-// Query holds the query string and the query parser.
+// A Query is the compiled form of a query.
 type Query struct {
-	str    string
-	parser *QueryParser
+	ast   syntax.Query
```

Comment:
why `ast`?

Reply:
To be clear at the usage site about the distinction between query-as-text vs. query-as-structure. We have a lot of code that uses "query" interchangeably for all of those, and I keep getting confused by it. 🙂